rules/limitdatetime/classes/rule.php

```diff
@@ -1,54 +1,99 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace registrationrule_limitdatetime;
+
+use coding_exception;
+use MoodleQuickForm;
+use stdClass;
+use tool_registrationrules\local\rule\configurable;
+use tool_registrationrules\local\rule_check_result;
+
+/**
+ * Reference implementation of a registration rule subplugin.
+ *
+ * @package   registrationrule_limitdatetime
+ * @copyright 2024 Catalyst IT Europe {@link https://www.catalyst-eu.net}
+ *            2024 eDaktik GmbH {@link https://www.edaktik.at/}
+ *            2024 lern.link GmbH {@link https://lern.link/}
+ *            2024 University of Strathclyde {@link https://www.strath.ac.uk}
+ * @author    Philipp Hager <philipp.hager@edaktik.at>
+ * @author    Lukas MuLu Müller <info@mulu.at>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class rule extends \tool_registrationrules\local\rule\rule_base implements configurable {
+    /** @var stdClass rule instance configuration */
+    private stdClass $config;
+
+    /** Names of fields added to the rule's settings form */
+    const SETTINGS_FIELDS = ['limitdatetime_from', 'limitdatetime_to'];
+
+    /**
+     * Constructor
+     *
+     * @param stdClass $config rule instance configuration
+     */
+    public function __construct($config) {
+        $this->config = $config;
+        parent::__construct($config);
+    }
+
+    /**
+     * Inject rule type specific settings into basic rule settings form if the type needs additional configuration.
+     *
+     * @param MoodleQuickForm $mform
+     * @return void
+     * @throws coding_exception
+     */
+    public static function extend_settings_form(MoodleQuickForm $mform): void {
+        $mform->addElement('date_time_selector', 'limitdatetime_from', get_string('from'));
+
+        $mform->addElement('date_time_selector', 'limitdatetime_to', get_string('to'));
+    }
+
+    /**
+     * Perform rule's checks applicable without any user input before the signup form is displayed.
+     *
+     * @return rule_check_result|null A rule_check_result object or null if check not applicable for this type.
+     */
+    public function pre_data_check(): ?rule_check_result {
+        $now = time();
+
+        /* A wizard is never late, nor is he early,
+         * he arrives precisely when he means to.
+         *
+         * Just like users don't.
+         * TODO: check timezone used in settings and maybe explain about used timezone as hint in UI?
```

Comment:
Oh gosh, are we limiting this to the server timezone, offsetting it to the site's timezone, head explodes!

Reply:
We currently use Moodle forms API's `date_time_selector` Elements without explicit further configuration, defaulting to the server timezone.

It may be best if we also inject a descriptive text explaining that the server timezone is used and may hinting to the currently viewing users timezone if it differs from the server timezone.
That way every user configuring the rule could calculate the expected times correctly.

See: https://docs.moodle.org/dev/lib/formslib.php_Form_Definition#date_time_selector and https://github.com/daledavies/moodle-tool_registrationrules/pull/17/files/610a8c4d8f7b0a5daa5562f68a04ffdc377d3187#diff-25c3fcf165260c54f5428c0ed5824cdfbd6d74f405684ce1dc2ce1461048894eR61

Reply:
Let's raise a separate issue for this then as it requires some thought. We should do that for each of the TODO items.